File: scala/sourcecode-async/src/main/scala/com/tersesystems/echopraxia/scala/async/DefaultAsyncLoggerMethods.scala
package com.tersesystems.echopraxia.scala.async

import com.tersesystems.echopraxia.api.Level._
import com.tersesystems.echopraxia.api.{Field, FieldBuilderResult, LoggerHandle, Value}
import com.tersesystems.echopraxia.scala.api.{Condition, DefaultMethodsSupport}
import sourcecode.{Enclosing, File, Line}

import java.util.function.Consumer

import scala.compat.java8.FunctionConverters._

/**
 * Default Async Logger Methods with source code implicits.
 */
trait DefaultAsyncLoggerMethods[FB] extends AsyncLoggerMethods[FB] {
  self: DefaultMethodsSupport[FB] =>

  protected def sourceInfoFields(
      fb: FB
  )(implicit line: Line, file: File, enc: Enclosing): FieldBuilderResult = {
    Field.keyValue(
      "sourcecode", // XXX make this configurable
      Value.`object`(
        Field.keyValue("file", Value.string(file.value)),
        Field.keyValue("line", Value.number(line.value)),
        Field.keyValue("enclosing", Value.string(enc.value))
      )
    )
  }

  /**
   * Logs using a logger handle at TRACE level.
   *
   * @param consumer the consumer of the logger handle.
   */
  // ------------------------------------------------------------------------
  // TRACE
  override def trace(consumer: LoggerHandle[FB] => Unit)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(TRACE, consumer.asJava, fieldBuilder)
  }

  /**
   * Logs using a condition and a logger handle at TRACE level.
   *
   * @param c        the condition
   * @param consumer the consumer of the logger handle.
   */
  override def trace(c: Condition, consumer: LoggerHandle[FB] => Unit)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(TRACE, c.asJava, consumer.asJava, fieldBuilder)
  }

  /**
   * Logs statement at TRACE level.
   *
   * @param message the given message.
   */
  override def trace(message: String)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(TRACE, (h: LoggerHandle[FB]) => h.log(message), fieldBuilder)
  }

  /**
   * Logs statement at TRACE level using a field builder function.
   *
   * @param message the message.
   * @param f       the field builder function.
   */
  override def trace(message: String, f: FB => FieldBuilderResult)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(TRACE, toConsumer(message, f), fieldBuilder)
  }

  /**
   * Logs statement at TRACE level with exception.
   *
   * @param message the message.
   * @param e       the given exception.
   */
  override def trace(message: String, e: Throwable)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(
        TRACE,
        (h: LoggerHandle[FB]) => h.log(message, (fb: FB) => onlyException(e)),
        fieldBuilder
      )
  }

  /**
   * Conditionally logs statement at TRACE level.
   *
   * @param condition the given condition.
   * @param message   the message.
   */
  override def trace(condition: Condition, message: String)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(TRACE, condition.asJava, (h: LoggerHandle[FB]) => h.log(message), fieldBuilder)
  }

  /**
   * Conditionally logs statement at TRACE level using a field builder function.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param f         the field builder function.
   */
  override def trace(condition: Condition, message: String, f: FB => FieldBuilderResult)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(TRACE, condition.asJava, toConsumer(message, f), fieldBuilder)
  }

  /**
   * Conditionally logs statement at TRACE level with exception.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param e         the given exception.
   */
  override def trace(condition: Condition, message: String, e: Throwable)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(
        TRACE,
        condition.asJava,
        (h: LoggerHandle[FB]) => h.log(message, (fb: FB) => onlyException(e)),
        fieldBuilder
      )
  }

  /**
   * Logs using a logger handle at DEBUG level.
   *
   * @param consumer the consumer of the logger handle.
   */
  // DEBUG
  override def debug(consumer: LoggerHandle[FB] => Unit)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(DEBUG, consumer.asJava, fieldBuilder)
  }

  /**
   * Logs using a condition and a logger handle at DEBUG level.
   *
   * @param c        the condition
   * @param consumer the consumer of the logger handle.
   */
  override def debug(c: Condition, consumer: LoggerHandle[FB] => Unit)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(DEBUG, c.asJava, consumer.asJava, fieldBuilder)
  }

  /**
   * Logs statement at DEBUG level.
   *
   * @param message the given message.
   */
  override def debug(message: String)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(DEBUG, (h: LoggerHandle[FB]) => h.log(message), fieldBuilder)
  }

  /**
   * Logs statement at DEBUG level using a field builder function.
   *
   * @param message the message.
   * @param f       the field builder function.
   */
  override def debug(message: String, f: FB => FieldBuilderResult)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(DEBUG, toConsumer(message, f), fieldBuilder)
  }

  /**
   * Logs statement at DEBUG level with exception.
   *
   * @param message the message.
   * @param e       the given exception.
   */
  override def debug(message: String, e: Throwable)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(
        DEBUG,
        (h: LoggerHandle[FB]) => h.log(message, (fb: FB) => onlyException(e)),
        fieldBuilder
      )
  }

  /**
   * Conditionally logs statement at DEBUG level.
   *
   * @param condition the given condition.
   * @param message   the message.
   */
  override def debug(condition: Condition, message: String)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(DEBUG, condition.asJava, (h: LoggerHandle[FB]) => h.log(message), fieldBuilder)
  }

  /**
   * Conditionally logs statement at DEBUG level using a field builder function.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param f         the field builder function.
   */
  override def debug(condition: Condition, message: String, f: FB => FieldBuilderResult)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(DEBUG, condition.asJava, toConsumer(message, f), fieldBuilder)
  }

  /**
   * Conditionally logs statement at DEBUG level with exception.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param e         the given exception.
   */
  override def debug(condition: Condition, message: String, e: Throwable)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(
        DEBUG,
        condition.asJava,
        (h: LoggerHandle[FB]) => h.log(message, (fb: FB) => onlyException(e)),
        fieldBuilder
      )
  }

  /**
   * Logs using a logger handle at INFO level.
   *
   * @param consumer the consumer of the logger handle.
   */
  // INFO
  override def info(consumer: LoggerHandle[FB] => Unit)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(INFO, consumer.asJava, fieldBuilder)
  }

  /**
   * Logs using a condition and a logger handle at INFO level.
   *
   * @param c        the condition
   * @param consumer the consumer of the logger handle.
   */
  override def info(c: Condition, consumer: LoggerHandle[FB] => Unit)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(INFO, c.asJava, consumer.asJava, fieldBuilder)
  }

  /**
   * Logs statement at INFO level.
   *
   * @param message the given message.
   */
  override def info(message: String)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(INFO, (h: LoggerHandle[FB]) => h.log(message), fieldBuilder)
  }

  /**
   * Logs statement at INFO level using a field builder function.
   *
   * @param message the message.
   * @param f       the field builder function.
   */
  override def info(message: String, f: FB => FieldBuilderResult)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(INFO, toConsumer(message, f), fieldBuilder)
  }

  /**
   * Logs statement at INFO level with exception.
   *
   * @param message the message.
   * @param e       the given exception.
   */
  override def info(message: String, e: Throwable)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(
        INFO,
        (h: LoggerHandle[FB]) => h.log(message, (fb: FB) => onlyException(e)),
        fieldBuilder
      )
  }

  /**
   * Conditionally logs statement at INFO level.
   *
   * @param condition the given condition.
   * @param message   the message.
   */
  override def info(condition: Condition, message: String)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(INFO, condition.asJava, (h: LoggerHandle[FB]) => h.log(message), fieldBuilder)
  }

  /**
   * Conditionally logs statement at INFO level using a field builder function.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param f         the field builder function.
   */
  override def info(condition: Condition, message: String, f: FB => FieldBuilderResult)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(INFO, condition.asJava, toConsumer(message, f), fieldBuilder)
  }

  /**
   * Conditionally logs statement at INFO level with exception.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param e         the given exception.
   */
  override def info(condition: Condition, message: String, e: Throwable)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(
        INFO,
        condition.asJava,
        (h: LoggerHandle[FB]) => h.log(message, (fb: FB) => onlyException(e)),
        fieldBuilder
      )
  }

  /**
   * Logs using a logger handle at WARN level.
   *
   * @param consumer the consumer of the logger handle.
   */
  // WARN
  override def warn(consumer: LoggerHandle[FB] => Unit)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(WARN, consumer.asJava, fieldBuilder)
  }

  /**
   * Logs using a condition and a logger handle at WARN level.
   *
   * @param c        the condition
   * @param consumer the consumer of the logger handle.
   */
  override def warn(c: Condition, consumer: LoggerHandle[FB] => Unit)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(WARN, c.asJava, consumer.asJava, fieldBuilder)
  }

  /**
   * Logs statement at WARN level.
   *
   * @param message the given message.
   */
  override def warn(message: String)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(WARN, (h: LoggerHandle[FB]) => h.log(message), fieldBuilder)
  }

  /**
   * Logs statement at WARN level using a field builder function.
   *
   * @param message the message.
   * @param f       the field builder function.
   */
  override def warn(message: String, f: FB => FieldBuilderResult)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(WARN, toConsumer(message, f), fieldBuilder)
  }

  /**
   * Logs statement at WARN level with exception.
   *
   * @param message the message.
   * @param e       the given exception.
   */
  override def warn(message: String, e: Throwable)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(
        WARN,
        (h: LoggerHandle[FB]) => h.log(message, (fb: FB) => onlyException(e)),
        fieldBuilder
      )
  }

  override def warn(condition: Condition, message: String)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(WARN, condition.asJava, (h: LoggerHandle[FB]) => h.log(message), fieldBuilder)
  }

  override def warn(condition: Condition, message: String, f: FB => FieldBuilderResult)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(WARN, condition.asJava, toConsumer(message, f), fieldBuilder)
  }

  override def warn(condition: Condition, message: String, e: Throwable)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(
        WARN,
        condition.asJava,
        (h: LoggerHandle[FB]) => h.log(message, (fb: FB) => onlyException(e)),
        fieldBuilder
      )
  }

  /**
   * Logs using a logger handle at ERROR level.
   *
   * @param consumer the consumer of the logger handle.
   */
  // ERROR
  override def error(consumer: LoggerHandle[FB] => Unit)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(ERROR, consumer.asJava, fieldBuilder)
  }

  /**
   * Logs using a condition and a logger handle at ERROR level.
   *
   * @param c        the condition
   * @param consumer the consumer of the logger handle.
   */
  override def error(c: Condition, consumer: LoggerHandle[FB] => Unit)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(ERROR, c.asJava, consumer.asJava, fieldBuilder)
  }

  override def error(message: String)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(ERROR, (h: LoggerHandle[FB]) => h.log(message), fieldBuilder)
  }

  override def error(message: String, f: FB => FieldBuilderResult)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(ERROR, toConsumer(message, f), fieldBuilder)
  }

  override def error(message: String, e: Throwable)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(
        ERROR,
        (h: LoggerHandle[FB]) => h.log(message, (fb: FB) => onlyException(e)),
        fieldBuilder
      )
  }

  override def error(condition: Condition, message: String)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(ERROR, condition.asJava, (h: LoggerHandle[FB]) => h.log(message), fieldBuilder)
  }

  override def error(condition: Condition, message: String, f: FB => FieldBuilderResult)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(ERROR, condition.asJava, toConsumer(message, f), fieldBuilder)
  }

  override def error(condition: Condition, message: String, e: Throwable)(implicit
      line: sourcecode.Line,
      file: sourcecode.File,
      enc: sourcecode.Enclosing
  ): Unit = {
    core
      .withFields(sourceInfoFields, fieldBuilder)
      .asyncLog(
        ERROR,
        condition.asJava,
        (h: LoggerHandle[FB]) => h.log(message, (fb: FB) => onlyException(e)),
        fieldBuilder
      )
  }

  @inline
  private def toConsumer(
      message: String,
      f: FB => FieldBuilderResult
  ): Consumer[LoggerHandle[FB]] = h => h.log(message, f.asJava)

  private def onlyException(e: Throwable): FieldBuilderResult = {
    Field.keyValue(Field.EXCEPTION, Value.exception(e))
  }

}

File: async/src/main/java/echopraxia/async/DefaultAsyncLoggerMethods.java
package echopraxia.async;

import static echopraxia.api.FieldConstants.EXCEPTION;
import static echopraxia.api.Value.exception;
import static echopraxia.logging.api.Level.*;

import echopraxia.api.Field;
import echopraxia.api.FieldBuilderResult;
import echopraxia.logging.api.Condition;
import echopraxia.logging.api.LoggerHandle;
import echopraxia.logging.spi.DefaultMethodsSupport;
import java.util.function.Consumer;
import java.util.function.Function;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Async logger methods that implement the default `core.asyncLog` delegation.
 *
 * @param <FB> the field builder.
 */
public interface DefaultAsyncLoggerMethods<FB>
    extends AsyncLoggerMethods<FB>, DefaultMethodsSupport<FB> {

  // ------------------------------------------------------------------------
  // TRACE

  /**
   * Logs using a logger handle at TRACE level.
   *
   * @param consumer the consumer of the logger handle.
   */
  default void trace(@NotNull Consumer<LoggerHandle<FB>> consumer) {
    core().asyncLog(TRACE, consumer, fieldBuilder());
  }

  /**
   * Logs using a condition and a logger handle at TRACE level.
   *
   * @param c the condition
   * @param consumer the consumer of the logger handle.
   */
  default void trace(@NotNull Condition c, @NotNull Consumer<LoggerHandle<FB>> consumer) {
    core().asyncLog(TRACE, c, consumer, fieldBuilder());
  }

  /**
   * Logs statement at TRACE level.
   *
   * @param message the given message.
   */
  default void trace(@Nullable String message) {
    core().asyncLog(TRACE, h -> h.log(message), fieldBuilder());
  }

  /**
   * Logs statement at TRACE level using a field builder function.
   *
   * @param message the message.
   * @param f the field builder function.
   */
  default void trace(@Nullable String message, @NotNull Function<FB, FieldBuilderResult> f) {
    core().asyncLog(TRACE, (LoggerHandle<FB> h) -> h.log(message, f), fieldBuilder());
  }

  /**
   * Logs statement at TRACE level with exception.
   *
   * @param message the message.
   * @param e the given exception.
   */
  default void trace(@Nullable String message, @NotNull Throwable e) {
    core()
        .asyncLog(
            TRACE,
            h -> h.log(message, fb -> Field.keyValue(EXCEPTION, exception(e))),
            fieldBuilder());
  }

  /**
   * Conditionally logs statement at TRACE level.
   *
   * @param condition the given condition.
   * @param message the message.
   */
  default void trace(@NotNull Condition condition, @Nullable String message) {
    core().asyncLog(TRACE, condition, h -> h.log(message), fieldBuilder());
  }

  /**
   * Conditionally logs statement at TRACE level using a field builder function.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param f the field builder function.
   */
  default void trace(
      @NotNull Condition condition,
      @Nullable String message,
      @NotNull Function<FB, FieldBuilderResult> f) {
    core().asyncLog(TRACE, condition, (LoggerHandle<FB> h) -> h.log(message, f), fieldBuilder());
  }

  /**
   * Conditionally logs statement at TRACE level with exception.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param e the given exception.
   */
  default void trace(@NotNull Condition condition, @Nullable String message, @NotNull Throwable e) {
    core()
        .asyncLog(
            TRACE,
            condition,
            h -> h.log(message, fb -> Field.keyValue(EXCEPTION, exception(e))),
            fieldBuilder());
  }

  // ------------------------------------------------------------------------
  // DEBUG

  /**
   * Logs using a logger handle at DEBUG level.
   *
   * @param consumer the consumer of the logger handle.
   */
  default void debug(@NotNull Consumer<LoggerHandle<FB>> consumer) {
    core().asyncLog(DEBUG, consumer, fieldBuilder());
  }

  /**
   * Logs using a condition and a logger handle at DEBUG level.
   *
   * @param c the condition
   * @param consumer the consumer of the logger handle.
   */
  default void debug(@NotNull Condition c, @NotNull Consumer<LoggerHandle<FB>> consumer) {
    core().asyncLog(DEBUG, c, consumer, fieldBuilder());
  }

  /**
   * Logs statement at DEBUG level.
   *
   * @param message the given message.
   */
  default void debug(@Nullable String message) {
    core().asyncLog(DEBUG, h -> h.log(message), fieldBuilder());
  }

  /**
   * Logs statement at DEBUG level using a field builder function.
   *
   * @param message the message.
   * @param f the field builder function.
   */
  default void debug(@Nullable String message, @NotNull Function<FB, FieldBuilderResult> f) {
    core().asyncLog(DEBUG, (LoggerHandle<FB> h) -> h.log(message, f), fieldBuilder());
  }

  /**
   * Logs statement at DEBUG level with exception.
   *
   * @param message the message.
   * @param e the given exception.
   */
  default void debug(@Nullable String message, @NotNull Throwable e) {
    core()
        .asyncLog(
            DEBUG,
            h -> h.log(message, fb -> (Field.keyValue(EXCEPTION, exception(e)))),
            fieldBuilder());
  }

  /**
   * Conditionally logs statement at DEBUG level.
   *
   * @param condition the given condition.
   * @param message the message.
   */
  default void debug(@NotNull Condition condition, @Nullable String message) {
    core().asyncLog(DEBUG, condition, h -> h.log(message), fieldBuilder());
  }

  /**
   * Conditionally logs statement at DEBUG level using a field builder function.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param f the field builder function.
   */
  default void debug(
      @NotNull Condition condition,
      @Nullable String message,
      @NotNull Function<FB, FieldBuilderResult> f) {
    core().asyncLog(DEBUG, condition, (LoggerHandle<FB> h) -> h.log(message, f), fieldBuilder());
  }

  /**
   * Conditionally logs statement at DEBUG level with exception.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param e the given exception.
   */
  default void debug(@NotNull Condition condition, @Nullable String message, @NotNull Throwable e) {
    core()
        .asyncLog(
            DEBUG,
            condition,
            h -> h.log(message, fb -> (Field.keyValue(EXCEPTION, exception(e)))),
            fieldBuilder());
  }

  // ------------------------------------------------------------------------
  // INFO

  /**
   * Logs using a logger handle at INFO level.
   *
   * @param consumer the consumer of the logger handle.
   */
  default void info(@NotNull Consumer<LoggerHandle<FB>> consumer) {
    core().asyncLog(INFO, consumer, fieldBuilder());
  }

  /**
   * Logs using a condition and a logger handle at INFO level.
   *
   * @param c the condition
   * @param consumer the consumer of the logger handle.
   */
  default void info(@NotNull Condition c, @NotNull Consumer<LoggerHandle<FB>> consumer) {
    core().asyncLog(INFO, c, consumer, fieldBuilder());
  }

  /**
   * Logs statement at INFO level.
   *
   * @param message the given message.
   */
  default void info(@Nullable String message) {
    core().asyncLog(INFO, h -> h.log(message), fieldBuilder());
  }

  /**
   * Logs statement at INFO level using a field builder function.
   *
   * @param message the message.
   * @param f the field builder function.
   */
  default void info(@Nullable String message, @NotNull Function<FB, FieldBuilderResult> f) {
    core().asyncLog(INFO, (LoggerHandle<FB> h) -> h.log(message, f), fieldBuilder());
  }

  /**
   * Logs statement at INFO level with exception.
   *
   * @param message the message.
   * @param e the given exception.
   */
  default void info(@Nullable String message, @NotNull Throwable e) {
    core()
        .asyncLog(
            INFO,
            h -> h.log(message, fb -> (Field.keyValue(EXCEPTION, exception(e)))),
            fieldBuilder());
  }

  /**
   * Conditionally logs statement at INFO level.
   *
   * @param condition the given condition.
   * @param message the message.
   */
  default void info(@NotNull Condition condition, @Nullable String message) {
    core().asyncLog(INFO, condition, h -> h.log(message), fieldBuilder());
  }

  /**
   * Conditionally logs statement at INFO level using a field builder function.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param f the field builder function.
   */
  default void info(
      @NotNull Condition condition,
      @Nullable String message,
      @NotNull Function<FB, FieldBuilderResult> f) {
    core().asyncLog(INFO, condition, (LoggerHandle<FB> h) -> h.log(message, f), fieldBuilder());
  }

  /**
   * Conditionally logs statement at INFO level with exception.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param e the given exception.
   */
  default void info(@NotNull Condition condition, @Nullable String message, @NotNull Throwable e) {
    core()
        .asyncLog(
            INFO,
            condition,
            h -> h.log(message, fb -> (Field.keyValue(EXCEPTION, exception(e)))),
            fieldBuilder());
  }

  // ------------------------------------------------------------------------
  // WARN

  /**
   * Logs using a logger handle at WARN level.
   *
   * @param consumer the consumer of the logger handle.
   */
  default void warn(@NotNull Consumer<LoggerHandle<FB>> consumer) {
    core().asyncLog(WARN, consumer, fieldBuilder());
  }

  /**
   * Logs using a condition and a logger handle at WARN level.
   *
   * @param c the condition
   * @param consumer the consumer of the logger handle.
   */
  default void warn(@NotNull Condition c, @NotNull Consumer<LoggerHandle<FB>> consumer) {
    core().asyncLog(WARN, c, consumer, fieldBuilder());
  }

  /**
   * Logs statement at WARN level.
   *
   * @param message the given message.
   */
  default void warn(@Nullable String message) {
    core().asyncLog(WARN, h -> h.log(message), fieldBuilder());
  }

  /**
   * Logs statement at WARN level using a field builder function.
   *
   * @param message the message.
   * @param f the field builder function.
   */
  default void warn(@Nullable String message, @NotNull Function<FB, FieldBuilderResult> f) {
    core().asyncLog(WARN, (LoggerHandle<FB> h) -> h.log(message, f), fieldBuilder());
  }

  /**
   * Logs statement at WARN level with exception.
   *
   * @param message the message.
   * @param e the given exception.
   */
  default void warn(@Nullable String message, @NotNull Throwable e) {
    core()
        .asyncLog(
            WARN,
            h -> h.log(message, fb -> (Field.keyValue(EXCEPTION, exception(e)))),
            fieldBuilder());
  }

  /**
   * Conditionally logs statement at INFO level.
   *
   * @param condition the given condition.
   * @param message the message.
   */
  default void warn(@NotNull Condition condition, @Nullable String message) {
    core().asyncLog(WARN, condition, h -> h.log(message), fieldBuilder());
  }

  /**
   * Conditionally logs statement at INFO level using a field builder function.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param f the field builder function.
   */
  default void warn(
      @NotNull Condition condition,
      @Nullable String message,
      @NotNull Function<FB, FieldBuilderResult> f) {
    core().asyncLog(WARN, condition, (LoggerHandle<FB> h) -> h.log(message, f), fieldBuilder());
  }

  /**
   * Conditionally logs statement at INFO level with exception.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param e the given exception.
   */
  default void warn(@NotNull Condition condition, @Nullable String message, @NotNull Throwable e) {
    core()
        .asyncLog(
            WARN,
            condition,
            h -> h.log(message, fb -> (Field.keyValue(EXCEPTION, exception(e)))),
            fieldBuilder());
  }

  // ------------------------------------------------------------------------
  // ERROR

  /**
   * Logs using a logger handle at ERROR level.
   *
   * @param consumer the consumer of the logger handle.
   */
  default void error(@NotNull Consumer<LoggerHandle<FB>> consumer) {
    core().asyncLog(ERROR, consumer, fieldBuilder());
  }

  /**
   * Logs using a condition and a logger handle at ERROR level.
   *
   * @param c the condition
   * @param consumer the consumer of the logger handle.
   */
  default void error(@NotNull Condition c, @NotNull Consumer<LoggerHandle<FB>> consumer) {
    core().asyncLog(ERROR, c, consumer, fieldBuilder());
  }

  /**
   * Logs statement at INFO level.
   *
   * @param message the given message.
   */
  default void error(@Nullable String message) {
    core().asyncLog(ERROR, h -> h.log(message), fieldBuilder());
  }

  /**
   * Logs statement at INFO level using a field builder function.
   *
   * @param message the message.
   * @param f the field builder function.
   */
  default void error(@Nullable String message, @NotNull Function<FB, FieldBuilderResult> f) {
    core().asyncLog(ERROR, (LoggerHandle<FB> h) -> h.log(message, f), fieldBuilder());
  }

  /**
   * Logs statement at INFO level with exception.
   *
   * @param message the message.
   * @param e the given exception.
   */
  default void error(@Nullable String message, @NotNull Throwable e) {
    core()
        .asyncLog(
            ERROR,
            h -> h.log(message, fb -> (Field.keyValue(EXCEPTION, exception(e)))),
            fieldBuilder());
  }

  /**
   * Conditionally logs statement at INFO level.
   *
   * @param condition the given condition.
   * @param message the message.
   */
  default void error(@NotNull Condition condition, @Nullable String message) {
    core().asyncLog(ERROR, condition, h -> h.log(message), fieldBuilder());
  }

  /**
   * Conditionally logs statement at INFO level using a field builder function.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param f the field builder function.
   */
  default void error(
      @NotNull Condition condition,
      @Nullable String message,
      @NotNull Function<FB, FieldBuilderResult> f) {
    core().asyncLog(ERROR, condition, (LoggerHandle<FB> h) -> h.log(message, f), fieldBuilder());
  }

  /**
   * Conditionally logs statement at INFO level with exception.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param e the given exception.
   */
  default void error(@NotNull Condition condition, @Nullable String message, @NotNull Throwable e) {
    core()
        .asyncLog(
            ERROR,
            condition,
            h -> h.log(message, fb -> (Field.keyValue(EXCEPTION, exception(e)))),
            fieldBuilder());
  }
}
